lib/api.js

```diff
@@ -0,0 +1,153 @@
+import base from 'taskcluster-base';
+import common from './common';
+import slugid from 'slugid';
+import _ from 'lodash';
+
+/**
+ * Check to see if a resource is stale
+ **/
+function hasExpired(expires) {
+  let currentTime = new Date();
+  let expireTime = new Date(expires);
+  if (currentTime > expireTime) {
+    return true;
+  }
+  return false;
+};
+
+/** API end-point for version v1/
+ *
+ * In this API implementation we shall assume the following context:
+ * {
+ *   publisher:      // publisher from base.Exchanges
+ * }
+ */
+var api = new base.API({
+  title:        "TaskCluster Secrets API Documentation",
+  description: [
+    "The secrets service, typically available at",
+    "`tools.taskcluster.net`, is responsible for managing",
+    "secure data in TaskCluster."
+  ].join('\n')
+});
+
+// Export API
+module.exports = api;
+
+/** Define tasks */
+api.declare({
+  method:      'put',
+  route:       '/name/:name(*)',
+  deferAuth:   true,
+  name:        'set',
+  input:       common.SCHEMA_PREFIX_CONST + "secret.json#",
+  scopes:      [['secrets:set:<name>']],
+  title:       'Create New Secret',
+  description: 'Set a secret associated with some key.'
+}, async function(req, res) {
+    req.satisfies(req.params);
+    let {name} = req.params;
+    let {secret, expires} = req.body;
+    try {
+      await this.entity.create({
+        name:       name,
+        secret:     secret,
+        expires:    new Date(expires)
+      });
+    } catch(e) {
+      res.status(e.statusCode).send(e.code);
+      return;
+    }
+    res.status(204).send();
+});
+
+api.declare({
+  method:      'patch',
+  route:       '/name/:name(*)',
+  deferAuth:   true,
+  name:        'update',
+  input:       common.SCHEMA_PREFIX_CONST + "secret.json#",
+  scopes:      [['secrets:update:<name>']],
+  title:       'Update A Secret',
+  description: 'Update a secret associated with some key.'
+}, async function(req, res) {
+    req.satisfies(req.params);
+    let {name} = req.params;
+    let {secret, expires} = req.body;
+    try {
+      let item = await this.entity.load({name: name});
+      await item.modify(function() {
+        this.secret = secret;
+        this.expires = new Date(expires);
+      });
+    } catch(e) {
+      res.status(e.statusCode).send(e.code);
```

Comment:
Again don't expose internal errors... but maybe handle cases where the entity doesn't exist:
```js
let item = this.entity.load({name}, true);
if (!item) {
  return res.status(404).json(message: "Secret not found, hence, couldn't be updated.");
}
```
Easy hack `this.entity.load({...}, true);` returns null if the entity doesn't exist, rather than throwing  an error.

ES6, tip: writing `{name}` is the same as writing `{name: name}`.